src/rust/lib/shortcuts/src/lib.rs

```diff
@@ -0,0 +1,883 @@
+        if self.dirty {
+            self.dirty   = false;
+            self.dfa     = (&self.nfa).into();
+            self.pressed = default();
+        }
+    }
+}
+
+impl<T> Default for AutomataRegistryModel<T> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+
+
+// ========================
+// === AutomataRegistry ===
+// ========================
+
+// FIXME
+// Replace T with Vec<T> to prevent shortcut override.
+
+/// Shortcut registry implementation based on a finite state automata. When defining shortcuts,
+/// a nondeterministic finite automata (NFA) is created. On the first usage, is it being optimized
+/// to a deterministic finite automata (DFA), which provides a constant time to get results on every
+/// key stroke. This allows us to define very complex shortcuts (regexp-like) and keep high
+/// performance of their resolution. However, the first stage of this approach is costly, namely
+/// building the NFA and optimizing it to DFA. Please refer to the benchmarks in this library to
+/// compare between available implementations.
+#[derive(CloneRef,Debug,Derivative)]
+#[derivative(Clone(bound=""))]
+#[derivative(Default(bound=""))]
+pub struct AutomataRegistry<T> {
+    rc : Rc<RefCell<AutomataRegistryModel<T>>>
+}
+
+impl<T> AutomataRegistry<T> {
+    /// Constructor.
+    pub fn new() -> Self {
+        default()
+    }
+}
+
+impl<T:Clone> AutomataRegistry<T> {
+    /// Repesent the NFA graph as a GraphViz DOT code.
+    pub fn nfa_as_graphviz_code(&self) -> String {
+        self.rc.borrow().nfa.as_graphviz_code()
+    }
+
+    /// Repesent the DFA graph as a GraphViz DOT code.
+    pub fn dfa_as_graphviz_code(&self) -> String {
+        self.rc.borrow().dfa.as_graphviz_code()
+    }
+
+    /// Get the approximate memory consumption of this shortcut registry DFA network.
+    pub fn approx_dfa_memory_consumption_mb(&mut self) -> f32 {
+        self.rc.borrow_mut().approx_dfa_memory_consumption_mb()
+    }
+}
+
+impl<T:Clone> Registry<T> for AutomataRegistry<T> {
+    fn add(&self, action_type:ActionType, expr:impl AsRef<str>, action:impl Into<T>) {
+        self.rc.borrow_mut().add(action_type,expr,action)
+    }
+
+    /// Process the press input event. See `on_event` docs to learn more.
+    fn on_press(&self, input:impl AsRef<str>) -> Vec<T> {
+        self.rc.borrow_mut().on_press(input)
+    }
+
+    /// Process the release input event. See `on_event` docs to learn more.
+    fn on_release(&self, input:impl AsRef<str>) -> Vec<T> {
+        self.rc.borrow_mut().on_release(input)
+    }
+
+    fn optimize(&self) {
+        self.rc.borrow_mut().optimize();
+    }
+}
+
+
+
+// ============================
+// === HashSetRegistryModel ===
+// ============================
+
+pub trait HashSetRegistryItem = Clone+Debug+Eq+Hash;
+
+/// Internal model for `HashSetRegistry`.
+#[derive(Debug)]
+pub struct HashSetRegistryModel<T> {
+    actions       : HashMap<ActionType,HashMap<String,Vec<T>>>,
```

Comment:
Consider `HashMap<(ActionType,String),Vec<T>>`

Reply:
I implemented it now and I dont like it. It forces us to do a lot more clones, for example when lookuping the map, as you need then to provide `&(ActionType,String)`, instead of `&ActionType`, and `&String`.